### ISSUE TITLE
TypeError in osys.update_leaves()

### ISSUE THREAD

**Issue by Jashcraf** [LouisDesdoigts/dLux#123]
Hi there! New user to dLux and am trying to run the [Phase Retrieval in dLux](https://louisdesdoigts.github.io/dLux/notebooks/phase_retrieval_demo/) demo on Google Collab. 

I'm just copy-pasting the code into a collab notebook, but when it's time to run this block of code I get a TypeError from `osys_update_leaves()`
![image](https://user-images.githubusercontent.com/25557892/193335420-d19fa6e9-4703-433f-8837-1aeaecd5a4ef.png)

Could anyone point me to what might be going wrong?

**Comment by Jashcraf** [comment]
Got the same error in the [Phase mask design](https://louisdesdoigts.github.io/dLux/notebooks/designing_a_mask/) demo, I suspect this is a me problem and not a dLux problem.

![image](https://user-images.githubusercontent.com/25557892/193337713-8b84f6db-d431-432e-baeb-e07a199075ca.png)

**Comment by Jordan-Dennis** [comment]
Hey @Jashcraf,
This is a strange error :laughing: that I've not seen before. Would you mind sharing the `python` version that you are using (typing `python --version` in the terminal will work). I just ran the notebook up to that point on my machine with `python==3.10.4` and it seemed to work. `typing` the module causing the error is part of the standard library so we may need to include a `python` version in `requirements.txt`, if this is the problem. 

Thanks for the issue

Jordan

**Comment by Jashcraf** [comment]
Hi @Jordan-Dennis 👋 
I'm using python version 3.7.14 right now because I'm trying dLux out on google collab. I can try install 3.10.4 to see if I get the same error.

**Comment by Jashcraf** [comment]
Maybe I can't, it looks like Google Collab only supports [Python 3.7 and 3.6](https://github.com/googlecolab/colabtools/issues/2165 
)

Does dLux work on M1 chips? I tried importing Jax and it was a little rude to my poor computer.
![image](https://user-images.githubusercontent.com/25557892/193362963-7a615415-5115-4319-9973-59958a811d4d.png)

**Comment by Jordan-Dennis** [comment]
I just checked and it is limited to CPU is the only downside.

**Comment by Jashcraf** [comment]
Oh that's fine I'm not trying to do anything crazy, just optimize some thin films to make finding the aliens a little easier

I saw that I might need to build it separately https://github.com/google/jax/issues/5501#issuecomment-792238148

**Comment by Jordan-Dennis** [comment]
Yeah, (this)[https://stackoverflow.com/questions/70815864/how-to-install-trax-jax-jaxlib-on-m1-mac-on-macos-12] stack post links in a whole lot of resources but that is the best I can do I'm afraid. We might change our use of the `typing` module anyway if it will cause problems with Collab. Its mainly a synthetic thing and has a very minimal runtime type checker which you have run into obviously, :laughing:

**Comment by Jashcraf** [comment]
Long ago I accepted my curse of running into trouble installing packages every time I start a new project - I _will_ prevail.

Thanks for the help! I'll give this a shot :)

**Comment by Jashcraf** [comment]
Steps I used to build Jax from Source taken from [Jax's documentation](https://jax.readthedocs.io/en/latest/developer.html)

**Build jaxlib**
1) `git clone https://github.com/google/jax`
2) `cd jax`
3) Install Xcode & command line utils from App store
4) Realize that you need to update your OS
5) Play Assasin's Creed for 4 hours while your OS updates because it's a weekend and you deserve it
6) Install Xcode
7) `python build/build.py`
8) doesn't execute successfully. 
9) panic

Well let's try the method outlined in [this Stack Overflew thread](https://stackoverflow.com/questions/70815864/how-to-install-trax-jax-jaxlib-on-m1-mac-on-macos-12).
1) Create new conda environment
2) `conda install numpy; conda install six`
3) `pip install -U pip pip install -U https://storage.googleapis.com/jax-releases/mac/jaxlib-0.1.75-cp310-none-macosx_11_0_arm64.whl`
4) Not a supported wheel on this platform, panic again
5) `python build/build.py --target_cpu darwin_arm64`
6) Doesn't build successfully
7) `pip install --upgrade "jax[cpu]"`
8) Doesn't build successfully

Going to try coming back to this again later.

**Comment by benjaminpope** [comment]
Try installing from conda forge.

conda install -c conda-forge jaxlib
conda install -c conda-forge jax

Works fine on my M1 and my older mac

———————————————
Dr Benjamin Pope (he/him)
Lecturer in Astrophysics
University of Queensland
benjaminpope.github.io
________________________________

**Comment by Jordan-Dennis** [comment]
I know that it is a bit late, but is there any chance you could repeat the error. I'm curious as to the 4 hidden frames since we don't actually use generic types. This seem's like something that might come from `equinox` perhaps and if it is you might be better chatting to @patrick-kidger from (equinox)[https://github.com/patrick-kidger/equinox/tree/main/equinox].

**Comment by Jashcraf** [comment]
@benjaminpope trying to install from conda forge with a fresh environment. Had to downgrade to < numpy v1.23, so I switched to 1.22. 

Tried `import jax` and it raised an ImportError.

Got an error that I'm having trouble understanding.  
<img width="988" alt="image" src="https://user-images.githubusercontent.com/25557892/193498689-a61790b6-0e7d-448c-afc0-5e722385bb95.png">

If I run the cell again it raises an AttributeError
<img width="987" alt="image" src="https://user-images.githubusercontent.com/25557892/193498843-c9e6c9a7-4200-4776-8671-6ccaf99047da.png">

I see that I didn't install Python 3.10.4 like @Jordan-Dennis recommended, so I'm upgrading using `conda install python=3.10.4`. During the installation there were some apparent conflicts with jax, so I'm just going to try start from the top and re-install jax.

Same errors unfortunately with `import jax` in python 3.10.4. 

`import jaxlib` seems to work just fine still.

**Comment by benjaminpope** [comment]
Let us know how that goes... the error messages indicate you're using Python 3.8 which does seem like the culprit. I would suggest creating a conda environment with Python 3.10, and using conda-forge to install jaxlib then jax

**Comment by Jashcraf** [comment]
Here's something weird (that's probably just my conda inexperience), when I create a new conda version and install python 3.10 checking the version verifies that I have the right installation
<img width="375" alt="image" src="https://user-images.githubusercontent.com/25557892/193600375-f003379e-2958-4258-bec1-8641b77dcb18.png">
Installing jax and jaxlib from conda-forge work just fine, but after the install when I try to open python to test the jax import, it shows as python version 3.8
<img width="598" alt="image" src="https://user-images.githubusercontent.com/25557892/193600625-60d4f061-8437-482b-938b-8b3e003a5fe2.png">

Not sure where in the pipeline the python version is changing, but I'll keep looking

**Comment by Jashcraf** [comment]
Installing from pip with a fresh python 3.10 conda environment doesn't change the python version, but we return to the error where my CPU and OS are not supported.
<img width="711" alt="image" src="https://user-images.githubusercontent.com/25557892/193603710-c9b454d8-1f01-42c8-9b45-b5421377eb2e.png">

**Comment by Jashcraf** [comment]
This is turning into maybe a problem where I can't install Jax, rather than being a dLux problem, so maybe it's worth closing the issue so I stop bugging you all?

**Comment by Jordan-Dennis** [comment]
Hahaha, I don't mind, we like having users and these are things that we need to think about. If you are able to work it out we will probably use it in an installation troubleshooting guide so it is helpful to us. I would be a little scared of the `Mlir` errors. This is the intermediate language representation that google uses to compile XLA. Running from `ubuntu` which is not so different from `MacOS` I am able to just run `pip install jax` and `pip install jaxlib` without any problems in a `conda environment`, although I think `pip install jax[cpu]` might be the recommended one.

**Comment by Jashcraf** [comment]
So I also have a windows machine, and somehow I've gotten it working on that using this method of installing jax https://github.com/cloudhan/jax-windows-builder

The phase mask design example executed like a charm!
![image](https://user-images.githubusercontent.com/25557892/193691981-363121da-d688-42e0-847d-44dd5468f5fe.png)

I'll continue sleuthing tomorrow, or maybe open an issue on Jax's github to see if they can figure out what might be going wrong.

**Comment by Jordan-Dennis** [comment]
Haha, that was a process, good job :+1:. I'm super busy at the moment but I want to look into your original problem with the `typing` module and find out if it was our usage or `equinox`'s that caused the problem. If it is a problem with us then we might just ditch `typing` in the relevant place. If it is a problem with `equinox` I will raise and issue with @patrick-kidger.

**Comment by LouisDesdoigts** [comment]
So I suspect that @Jashcraf may be running this through a Rosetta version of bash/zsh. I can't see why else it would possibly give a not supported error for the ARM wheel when running on an ARM architecture.

As for the typing, I think this may have been due to running the pip version of ∂Lux with the new tutorials. I could be wrong, I'll need to try and reproduce the error myself (unless @Jashcraf has the notebook handy)

**Comment by benjaminpope** [comment]
Time to push an update to PyPI?

**Comment by LouisDesdoigts** [comment]
Its always time to push an update to PyPi, there's too much dev happening 😂.

Next PyPi update is the 1.0, that's my goal.

**Comment by benjaminpope** [comment]
I'd probably push a version a bit sooner than that if we're doing collaboration with Jaren et al

**Comment by LouisDesdoigts** [comment]
Will push a new version by the end of the week - The new source/instrument stuff will force another tutorial update anyway.
